### PR TITLE
Bump multus-cni version

### PR DESCRIPTION
**Pull request by ormergi**
This PR fixes [1] by bumping multus-cni dependency to latest version [v3.4.1](https://github.com/intel/multus-cni/releases/tag/v3.4.1), updating nesessery changes and addes unit tests to prevent bugs related to [1].

[1] https://bugzilla.redhat.com/show_bug.cgi?id=1816971

Signed-off-by: Or Mergi <ormergi@redhat.com>

### CODE REVIEW

**Comment by phoracek** [comment]
To make sure this indeed fixes the issue we had with multiple IPs, could you please add a test requesting array of ips: https://bugzilla.redhat.com/show_bug.cgi?id=1816971

Since this is for Pods, unit test would be sufficient.

**Comment by ormergi** [comment]
/hold
should be merged after #126

File: pkg/pool-manager/pool_test.go [hunk 1]
```diff
@@ -389,7 +389,7 @@ var _ = Describe("Pool", func() {
 			_, exist = poolManager.macPoolMap["02:00:00:00:00:01"]
 			Expect(exist).To(BeTrue())
 
-			Expect(newPod.Annotations[networksAnnotation]).To(Equal(`[{"name":"ovs-conf","namespace":"default","mac":"02:00:00:00:00:01"}]`))
+			Expect(newPod.Annotations[networksAnnotation]).To(Equal(`[{"name":"ovs-conf","namespace":"default","mac":"02:00:00:00:00:01","cni-args":null}]`))
```
Comment (RamLavi) [inline pkg/pool-manager/pool_test.go:393]: small typo in commit message: `on of them is CNIArgs [2]` on --> one

Comment (ormergi) [reply pkg/pool-manager/pool_test.go:393]: fixed

File: pkg/controller/pod/pod_controller.go [hunk 2]
```diff
@@ -82,7 +82,7 @@ func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result
 		if errors.IsNotFound(err) {
 			err := r.poolManager.ReleasePodMac(fmt.Sprintf("%s/%s", request.Namespace, request.Name))
 			if err != nil {
-				logging.Printf(logging.ErrorLevel, "failed to release mac for pod %s: %v", request.NamespacedName, err)
+				logging.Errorf("%v, failed to release mac for pod %s: %v",logging.ErrorLevel, request.NamespacedName, err)
```
Comment (phoracek) [inline pkg/controller/pod/pod_controller.go:85]: I don't get how is this related to bump of Multus. If it is really needed, please explain it in the commit message.

File: pkg/pool-manager/pool_test.go [hunk 3]
```diff
@@ -413,5 +414,53 @@ var _ = Describe("Pool", func() {
 			Expect(newPod.Annotations[networksAnnotation]).To(Equal(afterAllocationAnnotation[networksAnnotation]))
 		})
 	})
+	Describe("Multus Network Annotations API Tests", func() {
+		It("should allocate requested mac-address with single ip-address request in network annotation", func() {
+			poolManager := createPoolManager("02:00:00:00:00:00", "02:00:00:00:00:01", &vmConfigMap)
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:419]: I know this follows the existing tests, but let's try to make our tests better today. Setup of the environment should be done in Context by BeforeEach. So in this case:

```go
Describe("... API Tests",... {

    Context("when pool is configured with two available addresses", ... {
        BeforeEach({
            // the pool setup
        })

        Context("and a Pod is created with single IP requested",...
            BeforeEach({
                // the VM setup
            })

            It("should successfullly ....", {
                // actual test
            })
    })
})

File: pkg/pool-manager/pool_test.go [hunk 4]
```diff
@@ -413,5 +414,53 @@ var _ = Describe("Pool", func() {
...
+			Expect(poolManager).ToNot(Equal(nil), "failed to create pool-manager")
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:420]: The expectation string should be what **should** happen. So in this case "Should create pool-manager". If we align to this rule everywhere in the code, then we can just read the "strings" in the tests to understand what is going on.

File: pkg/pool-manager/pool_test.go [hunk 5]
```diff
@@ -413,5 +414,53 @@ var _ = Describe("Pool", func() {
...
 
+			expectedMacRequestWithIpAnnotation := `[{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1"],"mac":"02:00:00:00:00:00","cni-args":null}]`
+			newPod := samplePod
+			newPod.Name = "testPod"
+			newPod.Annotations = map[string]string{networksAnnotation: `[{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1"]}]`}
+
+			err := poolManager.AllocatePodMac(&newPod)
+			Expect(err).ToNot(HaveOccurred(),"failed to handle allocate-pod-mac request with ip-address")
+			Expect(newPod.Annotations[networksAnnotation]).To(Equal(expectedMacRequestWithIpAnnotation),
+				fmt.Sprintf("Pod networtk annotation is not as expected,\n"+
```
Comment (qinqon) [inline pkg/pool-manager/pool_test.go:430]: You have to use Gomega JSON matcher MatchJSON so this check is not affected by json attribute re-order.

The "obtained" and "expected" typical test log has to be printed out by Gomega not by us, the string should be something like "Should annotate pod with IP and MAC"

File: pkg/pool-manager/pool_test.go [hunk 6]
```diff
@@ -413,5 +414,53 @@ var _ = Describe("Pool", func() {
...
+
+			expectedMacRequestWithIpsAnnotation := `[{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1","10.10.0.2","10.0.0.3"],"mac":"02:00:00:00:00:00","cni-args":null}]`
+			newPod := samplePod
+			newPod.Name = "testPod"
+			newPod.Annotations = map[string]string{networksAnnotation: `[{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1","10.10.0.2","10.0.0.3"]}]`}
+
+			err := poolManager.AllocatePodMac(&newPod)
+			Expect(err).ToNot(HaveOccurred(), "failed to handle allocate-pod-mac request with multiple ip-addresses")
+			Expect(newPod.Annotations[networksAnnotation]).To(Equal(expectedMacRequestWithIpsAnnotation),
+				fmt.Sprintf("Pod networtk annotation is not as expected,\n"+
```
Comment (qinqon) [inline pkg/pool-manager/pool_test.go:448]: ditto

File: pkg/controller/pod/pod_controller.go (hunk 2 above)
Comment (ormergi) [reply pkg/controller/pod/pod_controller.go:85]: This function so no longer public since this PR https://github.com/intel/multus-cni/pull/339
I updated the commit message.

Comment (phoracek) [reply pkg/controller/pod/pod_controller.go:85]: Ack

File: pkg/pool-manager/pool_test.go (hunk 4 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:420]: For me it is more clear to see these kind of error messages in after test suite logs,
specifying what failed and why instead of what should have happen.
I will align with this kind of error messages :).

Comment (phoracek) [reply pkg/pool-manager/pool_test.go:420]: Hopefully, you will read the code more often than seeing it fail. The positive string helps you read the code alone.

Here is an example :

```go
			It("should return a Result with RequeueAfter set (trigger re-reconciliation)", func() {
				result, err := reconciler.Reconcile(request)
				Expect(err).ToNot(HaveOccurred(), "Should succcessfully reconcile the request")
				Expect(result).To(Equal(reconcile.Result{RequeueAfter: 666 * time.Second}), "Result should have expected requeue value")
			})
```

The result (written in red is):

```
  Result should have expected requeue value
  Expected
      <reconcile.Result>: {
          Requeue: false,
          RequeueAfter: 5000000000,
      }
  to equal
      <reconcile.Result>: {
          Requeue: false,
          RequeueAfter: 666000000000,
      }
```

So you see the expectation and the actual result.

The biggest benefit of this is, that the code is way more readable when you just follow strings and don't need to revert their meaning to see what **should** happen.

File: pkg/pool-manager/pool_test.go (hunk 3 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:419]: ack

File: pkg/pool-manager/pool_test.go [hunk 7]
```diff
@@ -17,6 +17,8 @@ limitations under the License.
 package pool_manager
 
 import (
+	multus "github.com/intel/multus-cni/types"
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:20]: We usually split imports to 3 groups (sometimes 4 on test modules). The order is:

1. Built-in libraries ("net")
2. Third-party imports ("k8s.io/api/core/v1")
3. Internal imports ("github.com/k8snetworkplumbingwg/kubemacpool/pkg/names")

And then we sometimes plug a separate group with the test imports ("github.com/onsi/gomega") somewhere, but I can't really say we have a specific place for it.

By keeping this order, it is way easier for a reader of the code to quickly figure out what are the dependencies.

File: pkg/pool-manager/pool_test.go [hunk 8]
```diff
@@ -17,6 +17,8 @@ limitations under the License.
...
+	"github.com/k8snetworkplumbingwg/kubemacpool/build/_output/bin/go/src/encoding/json"
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:21]: We should not import generated code from "_output". That directory is not a part of the distribution. Your IDE probably automatically loaded a wrong lib.

File: pkg/pool-manager/pool_test.go [hunk 9]
```diff
@@ -414,4 +416,75 @@ var _ = Describe("Pool", func() {
 		})
 	})
 
-})
+	Describe("Multus Network Annotations API Tests", func() {
+		BeforeEach(func() {
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:420]: As noted offline, the before each should keep only the two following lines. poolManager needs to be initialized in the root of Describe and then populated in BeforeEach (we want to reset the manager per each test).

File: pkg/pool-manager/pool_test.go [hunk 10]
```diff
@@ -414,4 +416,75 @@ var _ = Describe("Pool", func() {
...
+			poolManager := createPoolManager("02:00:00:00:00:00", "02:00:00:00:00:01", &vmConfigMap)
+			Expect(poolManager).ToNot(Equal(nil), "should create pool-manager")
+
+			Context("When pool-manager is configured with available addresses", func(){
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:424]: Context without any content (except other contexts) is suspicions. If you are not setting anything, you are probably keeping the context implicit and defined in internal code. That is risky since it may get outdated quickly, it does not keep the context info visible and does not force us to actually follow the context.

I think that in this case, this Context is redundant since you describe the addresses parameter in the following Context. Please drop it.

Comment (phoracek) [inline pkg/pool-manager/pool_test.go:424]: For the `"When`. I think it is better to start with either lower case (so the test name is readable as one long sentence) or with all capitals `WHEN` (so it is CLEAR in the test name that the context starts there)

File: pkg/pool-manager/pool_test.go (hunk 7 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:20]: ack

File: pkg/pool-manager/pool_test.go (hunk 8 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:21]: yep, It was the IDE 
fixed

File: pkg/pool-manager/pool_test.go (hunk 9 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go]: Got it :)

File: pkg/pool-manager/pool_test.go (hunk 10 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go]: done

Comment (ormergi) [reply pkg/pool-manager/pool_test.go]: ack

File: pkg/pool-manager/pool_test.go (hunk 4 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go]: Thanks for explaining I will change it :)

File: pkg/pool-manager/pool_test.go (hunk 5 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go]: done

File: pkg/pool-manager/pool_test.go [hunk 11]
```diff
@@ -414,4 +416,61 @@ var _ = Describe("Pool", func() {
...
+						err = json.Unmarshal([]byte(obtainedNetworksAnnotationJson), &obtainedNetworksAnnotation)
+						Expect(err).ToNot(HaveOccurred(), "should convert obtained annotation as json to multus.NetworkSelectionElement")
+
+						By("Convert expected networks annotation JSON to multus.NetworkSelectionElement array")
+						expectedNetworksAnnotation := []multus.NetworkSelectionElement{}
+						err = json.Unmarshal([]byte(networkRequestAnnotation), &expectedNetworksAnnotation)
+						Expect(err).ToNot(HaveOccurred(), "should convert expected annotation as json to multus.NetworkSelectionElement")
+
+						By("Compare between each obtained and expected network request according to the mac-address")
+						for _, obtainedNetwork := range obtainedNetworksAnnotation {
```
Comment (qinqon) [inline pkg/pool-manager/pool_test.go:450]: Check if ContainElement or ContainElements can help you here https://onsi.github.io/gomega/#containelementselement-interface

File: pkg/pool-manager/pool_test.go [hunk 12]
```diff
@@ -414,4 +416,61 @@ var _ = Describe("Pool", func() {
...
+							for _, expectedNetwork := range expectedNetworksAnnotation {
```
Comment (qinqon) [inline pkg/pool-manager/pool_test.go:451]: Either check size before loop or convert slices into []string with the macRequest so you can use ContaineElements and it will already check the size for you.

File: pkg/pool-manager/pool_test.go (hunk 10 above)
Comment (phoracek) [reply pkg/pool-manager/pool_test.go]: @ormergi I don't see it being changed. There are still 2 contexts, the outter does not set anything.

File: pkg/pool-manager/pool_test.go [hunk 13]
```diff
@@ -414,4 +416,61 @@ var _ = Describe("Pool", func() {
 		})
 	})
 
+	Describe("Multus Network Annotations API Tests", func() {
+		pod := v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "testPod", Namespace: "default"}}
+		poolManager := &PoolManager{}
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:421]: I'm sorry, I commented this should be in `Describe`, while it should be in the `Context` where we set it in `BeforeEach`.

File: pkg/pool-manager/pool_test.go [hunk 14]
```diff
@@ -414,4 +416,61 @@ var _ = Describe("Pool", func() {
...
+						`[{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1","10.10.0.2","10.0.0.3"],"mac":"02:00:00:00:00:00"}]`),
+				)
+
+				table.DescribeTable("should fail to allocate requested mac-address", func(networkRequestAnnotation string) {
+					pod.Annotations = map[string]string{networksAnnotation: networkRequestAnnotation}
+					By("Request specific mac-address by adding the address to the networks pod annotation")
+					err := poolManager.AllocatePodMac(&pod)
+					Expect(err).To(HaveOccurred(), "should fail to allocate mac address due to bad annotation format")
+				},
+					table.Entry("with ip-address request as string instead of string array",
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:470]: No need for a describe table if there is only a single line. It extends the complexity and brings no value.

File: pkg/pool-manager/pool_test.go (hunk 11 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:450]: We need to bump Gomega to have ContainElements https://github.com/onsi/gomega/releases

File: pkg/pool-manager/pool_test.go (hunk 12 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:451]: I will add a break for the end of the if that should prevent overlapping, once we bump Gomega  we can use ContaineElements

**Comment by ormergi** [comment]
> Gomga ContainElement[s] can help here also there is a false possitive at one of the test.
> 
> I see the table test a little too much for too tests but I suppose now is easier to increase with more use cases right ?

I undersand why it may look too-much but I reached to a point I have two of the same test

File: pkg/pool-manager/pool_test.go (hunk 14 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:470]: I use DescribeTable here so I wont need to have the same exact tests with only one line change,where this line is actually the input for the test.
I added another case where we have multiple network requests, its much easier to do it like that.
Also you have the test logic is in once place so its easier to understand
what you think?

File: pkg/pool-manager/pool_test.go [hunk 15]
```diff
@@ -414,4 +416,56 @@ var _ = Describe("Pool", func() {
...
+				table.Entry("with single ip-address request as string array",
+				`[{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1"],"mac":"02:00:00:00:00:00"}]`),
+				table.Entry("with multiple ip-address request as string array",
+				`[{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1","10.10.0.2","10.0.0.3"],"mac":"02:00:00:00:00:00"}]`),
+				table.Entry("with multiple networks requsets",`[
+						{"name":"ovs-conf","namespace":"default","ips":["10.10.0.1","10.10.0.2","10.0.0.3"],"mac":"02:00:00:00:00:00"},
+						{"name":"cnv-bridge","namespace":"openshift-cnv","ips":["192.168.66.100","192.168.66.101"],"mac":"02:F0:F0:F0:F0:F0"}
+				]`),
+			)
+			It("should fail to allocate requested mac-address, with ip-address request as string instead of string array", func(){
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:462]: This is something that should be tested on Multus, but I don't see how it concerns us in KMP.

File: pkg/pool-manager/pool_test.go [hunk 16]
```diff
@@ -414,4 +416,56 @@ var _ = Describe("Pool", func() {
 		})
 	})
 
+	Describe("Multus Network Annotations API Tests", func() {
+		Context("when pool-manager is configured with available addresses, Pod is created with networks annotation with mac-address", func() {
```
Comment (phoracek) [inline pkg/pool-manager/pool_test.go:420]: This is missleading. You say that "Pod is created", but it is not created in the context (before each), it is created only in the describe table only. You can drop the variable from here and keep it in DescribeTable. You should also remove the "Pod is created" from this title.

**Comment by qinqon** [comment]
/lgtm
@phoracek you good to approve ?

File: pkg/pool-manager/pool_test.go (hunk 15 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go:462]: This is how we check that the pool-manager will not handle request that multus wouldn't that eventually will fail

Comment (phoracek) [reply pkg/pool-manager/pool_test.go:462]: If it will eventually fail on Multus, we should let it fail on Multus. There is a lot of other components that it may fail on later. We should not care about any of those. We should test for our failures and our positive flow. Duplicating tests of other components will leave us with unmaintainable code-base. Unless you convince me this brings any value to this project.

Comment (ormergi) [reply pkg/pool-manager/pool_test.go:462]: This PR is about fixing this bug https://bugzilla.redhat.com/show_bug.cgi?id=1816971
and preventing another issue like that, so I added two cases where we use the new `ips` property and another case where `ips` is not valid.
This is about testing [parsePodNetworkAnnotation](https://github.com/k8snetworkplumbingwg/kubemacpool/blob/97e9201d0609d2f3ae7f23a68b4893f5961f2113/pkg/pool-manager/pod_pool.go#L286) which is where the error comes from, if you think it is redundant I will remove it.

File: pkg/pool-manager/pool_test.go (hunk 16 above)
Comment (phoracek) [reply pkg/pool-manager/pool_test.go]: @ormergi you haven't dropped the `pod` initialization from the context. It is no longer needed there

File: pkg/pool-manager/pool_test.go (hunk 15 above)
Comment (phoracek) [reply pkg/pool-manager/pool_test.go:462]: I don't see how adding a specific negative test helps us. But maybe I'm missing something, I'd like to hear from the other reviewers.

@RamLavi @qinqon

Comment (qinqon) [reply pkg/pool-manager/pool_test.go:462]: Well testing the behaviour on bad format is not a bad thing, at least we know that we don't panic.

Comment (phoracek) [reply pkg/pool-manager/pool_test.go:462]: Fair enough then

Comment (ormergi) [reply pkg/pool-manager/pool_test.go]: If this test fail its indicate we are not aligned with the API of NetworkSelectionElement we parse and use

File: pkg/pool-manager/pool_test.go (hunk 16 above)
Comment (ormergi) [reply pkg/pool-manager/pool_test.go]: done

**Comment by phoracek** [comment]
Feel free to cancel the hold
